### PR TITLE
feat(related_issues): Trace connected errors

### LINKED ISSUE #69615: BE - ⏫ Trace-connected related issues
- [x] Investigate the trace connected API
- [x] API to list issue IDs of trace-connected errors 

Merged here: https://github.com/getsentry/sentry/pull/69237

### PR DESCRIPTION
**Pull request by armenzg**
Given a group, we look for the recommended event and search for any other errors in the trace.

These trace-connected groups will be shown under the Related Issues tab. See the current look (more UI changes will be needed).

image

### CODE REVIEW

File: tests/sentry/api/endpoints/issues/test_related_issues.py [hunk 1]
```diff
@@ -46,5 +46,19 @@ def test_same_root_related_issues(self) -> None:
...
+
+    def test_trace_connected_errors(self) -> None:
+        # This is the group we're going to query about
+        group = self.create_group(data=self._data(self.error_type, self.error_value))
+        self.group_id = group.id
+
+        response = self.get_success_response()
+        assert response.json() == {
+            "data": [
+                {"type": "same_root_cause", "data": []},
```
Comment (armenzg) [inline tests/sentry/api/endpoints/issues/test_related_issues.py:61]: This test is failing with this:
```
pytest -s -vv tests/sentry/api/endpoints/issues/test_related_issues.py
========================================================================= test session starts ==========================================================================
platform darwin -- Python 3.11.6, pytest-8.0.0, pluggy-1.4.0 -- /Users/armenzg/code/sentry/.venv/bin/python3
cachedir: .pytest_cache
metadata: {'Python': '3.11.6', 'Platform': 'macOS-14.3.1-arm64-arm-64bit', 'Packages': {'pytest': '8.0.0', 'pluggy': '1.4.0'}, 'Plugins': {'fail-slow': '0.3.0', 'rerunfailures': '11.0', 'pytest_sentry': '0.3.0rc1', 'json-report': '1.5.0', 'metadata': '3.1.1', 'time-machine': '2.13.0', 'xdist': '3.0.2', 'anyio': '3.7.1', 'django': '4.8.0', 'cov': '4.0.0'}}
django: version: 5.0.3
rootdir: /Users/armenzg/code/sentry
configfile: pyproject.toml
plugins: fail-slow-0.3.0, rerunfailures-11.0, pytest_sentry-0.3.0rc1, json-report-1.5.0, metadata-3.1.1, time-machine-2.13.0, xdist-3.0.2, anyio-3.7.1, django-4.8.0, cov-4.0.0
collected 2 items

tests/sentry/api/endpoints/issues/test_related_issues.py::RelatedIssuesTest::test_same_root_related_issues Creating test database for alias 'default' ('test_region')...
Creating test database for alias 'control' ('test_control')...
PASSED
tests/sentry/api/endpoints/issues/test_related_issues.py::RelatedIssuesTest::test_trace_connected_errors FAILEDDestroying test database for alias 'default' ('test_region')...
Destroying test database for alias 'control' ('test_control')...


=============================================================================== FAILURES ===============================================================================
____________________________________________________________ RelatedIssuesTest.test_trace_connected_errors _____________________________________________________________
tests/sentry/api/endpoints/issues/test_related_issues.py:59: in test_trace_connected_errors
    assert response.json() == {
E   AssertionError: assert {'data': [{'type': 'same_root_cause', 'data': [6]}, {'type': 'trace_connected', 'data': []}]} == {'data': [{'type': 'same_root_cause', 'data': []}, {'type': 'trace_connected', 'data': []}]}
E
E     Differing items:
E     {'data': [{'data': [6], 'type': 'same_root_cause'}, {'data': [], 'type': 'trace_connected'}]} != {'data': [{'data': [], 'type': 'same_root_cause'}, {'data': [], 'type': 'trace_connected'}]}
E
E     Full diff:
E       {
E           'data': [
E               {
E     -             'data': [],
E     ?                      --
E     +             'data': [
E     +                 6,
E     +             ],
E                   'type': 'same_root_cause',
E               },
E               {
E                   'data': [],
E                   'type': 'trace_connected',
E               },
E           ],
E       }
======================================================================= short test summary info ========================================================================
FAILED tests/sentry/api/endpoints/issues/test_related_issues.py::RelatedIssuesTest::test_trace_connected_errors - AssertionError: assert {'data': [{'type': 'same_root_cause', 'data': [6]}, {'type': 'trace_connected', 'data': []}]} == {'data': [{'type': 'same_root_cause', 'data...
===================================================================== 1 failed, 1 passed in 4.05s ======================================================================
```

File: src/sentry/api/endpoints/organization_events_trace.py [hunk 2]
```diff
@@ -543,28 +543,7 @@ def query_trace_data(
...
-            "tags[level]",
-        ],
-        # Don't add timestamp to this orderby as snuba will have to split the time range up and make multiple queries
-        orderby=["id"],
-        limit=limit,
-        config=QueryBuilderConfig(
-            auto_fields=False,
-        ),
-    )
+    error_query = find_errors_for_trace_id(params, trace_id, limit)
```
Comment (armenzg) [inline src/sentry/api/endpoints/organization_events_trace.py:546]: I'm going to call `find_errors_for_trace_id` from my endpoint, thus, making it its own function.

File: src/sentry/issues/related/same_root_cause.py [hunk 3]
```diff
@@ -11,7 +11,10 @@ def same_root_cause_analysis(group: Group) -> list[int]:
     """Analyze and create a group set if the group was caused by the same root cause."""
     # Querying the data field (which is a GzippedDictField) cannot be done via
     # Django's ORM, thus, we do so via compare_groups
-    project_groups = RangeQuerySetWrapper(Group.objects.filter(project=group.project_id), limit=100)
+    project_groups = RangeQuerySetWrapper(
+        Group.objects.filter(project=group.project_id).exclude(id=group.id),
```
Comment (armenzg) [inline src/sentry/issues/related/same_root_cause.py:15]: Unrelated change to exclude the group ID we're querying about.

This is the related test change for it:
image

File: src/sentry/search/events/builder/discover.py [hunk 4]
```diff
@@ -583,13 +583,16 @@ def resolve_params(self) -> list[WhereType]:
...
             conditions.append(Condition(self.column("timestamp"), Op.LT, self.end))
 
-        conditions.append(
-            Condition(
-                self.column("project_id"),
-                Op.IN,
-                self.params.project_ids,
+        # The if clause will prevent calling Snuba with an empty list of projects and complain with:
+        # sentry.utils.snuba.SnubaError: validation failed for entity events: missing required conditions for project_id
+        if self.params.project_ids:
```
Comment (armenzg) [inline src/sentry/search/events/builder/discover.py:588]: I put this change in #69499
I removed this change locally. I will push it tomorrow since I want to see all test results first.

File: src/sentry/testutils/factories.py [hunk 5]
```diff
@@ -1362,7 +1368,7 @@ def create_userreport(project, event_id=None, **kwargs):
         )
 
         return UserReport.objects.create(
-            group_id=event.group.id,
+            group_id=event.group_id,
```
Comment (armenzg) [inline src/sentry/testutils/factories.py:1371]: This works around a typing issue.

File: tests/sentry/api/endpoints/issues/test_related_issues.py [hunk 6]
```diff
@@ -1,39 +1,38 @@
...
 from django.urls import reverse
 
-from sentry.testutils.cases import APITestCase
+from sentry.testutils.cases import APITestCase, SnubaTestCase, TraceTestCase
 
 
-class RelatedIssuesTest(APITestCase):
+class RelatedIssuesTest(APITestCase, SnubaTestCase, TraceTestCase):
     endpoint = "sentry-api-0-issues-related-issues"
+    FEATURES: list[str] = []
```
Comment (armenzg) [inline tests/sentry/api/endpoints/issues/test_related_issues.py:8]: This is needed in one of the test cases. Ideally, I fix it up there.

File: tests/sentry/api/endpoints/issues/test_related_issues.py [hunk 7]
```diff
@@ -1,39 +1,38 @@
...
 
     def setUp(self) -> None:
         super().setUp()
         self.login_as(user=self.user)
         self.organization = self.create_organization(owner=self.user)
-        self.error_type = "ApiTimeoutError"
-        self.error_value = "Timed out attempting to reach host: api.github.com"
```
Comment (armenzg) [inline tests/sentry/api/endpoints/issues/test_related_issues.py:16]: Unrelated change. I could move it into its own PR.

File: src/sentry/api/endpoints/organization_events_trace.py (hunk 2 above)
Comment (wmak) [reply src/sentry/api/endpoints/organization_events_trace.py:546]: I'd prefer you re-implement this function rather than re-use here so that I can still make changes here without potentially breaking your endpoint. 

Especially because the query pattern here is very dependent on feeding the latter `augment_transactions_with_spans`

Comment (armenzg) [reply src/sentry/api/endpoints/organization_events_trace.py:546]: wfm

File: src/sentry/issues/related/trace_connected.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    query = QueryBuilder(
+        Dataset.Events,
+        {"start": start, "end": end, "organization_id": org_id, "project_id": project_ids},
+        query=f"trace:{event.trace_id}",
+        selected_columns=["id", "issue.id"],
+        # Don't add timestamp to this orderby as snuba will have to split the time range up and make multiple queries
+        orderby=["id"],
+        limit=100,
+        config=QueryBuilderConfig(auto_fields=False),
+    )
```
Comment (vartec) [inline src/sentry/issues/related/trace_connected.py:32]: this seems like potentially rather heavy query, maybe add some read-through cache?

**Review by vartec — APPROVED** [review]

File: src/sentry/issues/related/trace_connected.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+def trace_connected_analysis(group: Group) -> list[int]:
+    event = group.get_recommended_event_for_environments()
+    if not event or event.trace_id is None:
+        return []
+
+    org_id = group.project.organization_id
+    # XXX: Test without a list and validate the data type
+    project_ids = list(Project.objects.filter(organization_id=org_id).values_list("id", flat=True))
+    start, end = default_start_end_dates()  # Today to 90 days back
```
Comment (wmak) [inline src/sentry/issues/related/trace_connected.py:22]: +1 to Bartek's comment here, this is a giant query if its spanning multiple projects and ninety days, I'm not familiar with what you're building, but do you really need to scan 90d of data? Traces rarely span more than a few minutes, and even more rarely span multiple days.

That is to say, you have an event do you know that events timestamp? can you use that?

File: src/sentry/testutils/cases.py [hunk 10]
```diff
@@ -3469,8 +3469,8 @@ def populate_project1(self) -> None:
             )
         ]
 
-    def load_errors(self) -> tuple[Event, Event]:
-        """Generates 2 events for gen1 projects."""
+    def load_errors(self) -> tuple[Event, Event, Event]:
```
Comment (wmak) [inline src/sentry/testutils/cases.py:3472]: nit: if we're extending this to this point should we just return a list of events instead of a tuple?

File: src/sentry/issues/related/trace_connected.py (hunk 9 above)
Comment (armenzg) [reply src/sentry/issues/related/trace_connected.py:22]: I will follow up.
We can just use the timestamp of the event as the start date.

File: src/sentry/testutils/cases.py (hunk 10 above)
Comment (armenzg) [reply src/sentry/testutils/cases.py:3472]: Sure 👍🏻

File: src/sentry/issues/related/trace_connected.py (hunk 9 above)
Comment (mikejihbe) [reply src/sentry/issues/related/trace_connected.py:22]: You'd want to see the trace connected issues for downstream issues as well (e.g. the second issue in a trace). I think you could limit this safely to several hours either direction of the occurrence.
